test/BlockManager.js

```diff
@@ -516,5 +516,51 @@ describe('BlockManager', function () {
       assertBNEqual(dispute.lastVisitedStaker, sortedStakers[sortedStakers.length - 1], 'lastVisited should match');
       await blockManager.connect(signers[19]).finalizeDispute(epoch, 0);
     });
+    it('staker should not be able to propose when not elected', async function () {
+      await mineToNextEpoch();
+      const epoch = await getEpoch();
+      const stakerIdAcc8 = await stakeManager.stakerIds(signers[8].address);
+      const staker = await stakeManager.getStaker(stakerIdAcc8);
+      const { biggestInfluencerId } = await getBiggestInfluenceAndId(stakeManager);
+      const iteration = await getIteration(voteManager, stakeManager, staker);
+      const votes = [100, 200, 300, 400, 500, 600, 700, 800, 900];
+
+      const commitment1 = utils.solidityKeccak256(
+        ['uint32', 'uint48[]', 'bytes32'],
+        [epoch, votes, '0x727d5c9e6d18ed15ce7ac8d3cce6ec8a0e9c02481415c0823ea49d847ccb9ddd']
+      );
+      await voteManager.connect(signers[8]).commit(epoch, commitment1);
+      await mineToNextState();
+      await voteManager.connect(signers[8]).reveal(epoch, votes,
+        '0x727d5c9e6d18ed15ce7ac8d3cce6ec8a0e9c02481415c0823ea49d847ccb9ddd');
+      await mineToNextState();
+
+      const tx = blockManager.connect(signers[8]).propose(epoch,
+        [100, 200, 300, 400, 500, 600, 700, 800, 900],
+        iteration,
```

Comment:
hey 
how it's assured that staker won't be elected proposer here?

Reply:
it is because of 
if ((rand1 + 1) != stakerId) {
            return false;
        }

Reply:
oh cool so you are getting iterations in the previous epoch
so seed is secret of the last commit 
hence in the new epoch, previous interaction won't be true 
right?

Reply:
yes

Reply:
then there can be an issue 
there is a chance that the previous iteration would hold true for this epoch as well
failing this test case needlessly 
what could be a better approach 
is finding a invalid iteration in this epoch only

we can do so by adding the clone method of getIteration as getFirstInvalidIteration 
which would return the first iteration which won't hold the condition true.

Reply:
But How it can happen, the previous iteration being true for this epoch.

Reply:
because, for diff seeds, there can be the same iteration.
For example,
Your seed for this epoch is X and you find an iteration "i"
Then its not given that in the next epochs even if you change seed, the iteration cant be i
We find iteration from start only (0) everytime

Reply:
Ohkay, if I just do, iteration - 1 in the propose transaction, that can also help. because it gives the first iteration. It can work too?

Reply:
oh yes it will work for all cases except when iteration is 0

Reply:
okay

Reply:
@abhishekvispute @adi44  We can create a function in utils which finds the first Wrong Iteration like "getFirstFalseIteration" by this it will be sure that staker is not elected and also in that function we need to add if statement for 0 like if getIteration returns 0 we return 1 and then a for loop may be starting from 0.
We discussed this on call once I was doing that

Reply:
passing iteration-1 was creating problem for 0 and I think othe than 0 also I tried that it was returning true sometimes and false sometimes so we discussed to create a function.

Reply:
Other than zero, it should work as expected. For what iterations iteration - 1 was not working?

Reply:
It's not working for iteration = 2 also

Reply:
I checked now it's not working for 0 else it was working I checked 5 times. It's not working only for 0.

Reply:
But it's also working for 0 sometimes like 2 times out of 10